### PR TITLE
`cljr-slash` using `suggest-libspecs` middleware op

### PR DESCRIPTION
**Pull request by dgtized**
Implements https://github.com/clojure-emacs/clj-refactor.el/issues/531. ~~Still need to add changelog/readme updates if this approach is accepted.~~

If the defcustom `cljr-slash-uses-suggest-libspec` is enabled, `cljr-slash` will query the new `suggest-libspec` middleware op for suggested namespaces.

Currently it sends over the entire `cljr-magic-require-namespaces` as the preferred-aliases map, that may require pre-processing at some point, but the placeholder seems correct. It also correctly identifies if code block is in a conditional language context, and provides that information to the middleware to in order to assist generating recommendations.

Finally, as it's using `completing-read` to select namespaces, if the user inserts their own content into the response, that will be inserted as the new libspec. This allows the user to reference a new libspec, or edit it to include refers or require-macro. In the future, it might make sense to use the universal argument to temporarily force `cljr-magic-requires` to use :prompt, allowing a user to insert a require of their own choice more easily.

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests (if possible) to cover your change(s)
- [x] The new code is not generating byte compile warnings (run `cask exec emacs -batch -Q -L . -eval "(progn (setq byte-compile-error-on-warn t) (batch-byte-compile))" clj-refactor.el`)
- [x] All tests are passing (run `./run-tests.sh`)
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [x] You've updated the readme (if adding/changing user-visible functionality)
- [x] Fix wiki docs to reference the `cljr-slash-uses-suggest-libspec` defcustom

Thanks!

### CODE REVIEW

File: clj-refactor.el
Comment (vemv) [inline clj-refactor.el:78]:
```suggestion
  "If t, `cljr-slash' magic require functionality will use the newer
```

File: clj-refactor.el
Comment (vemv) [inline clj-refactor.el:1968]: Would seem nice to explain what both things are

File: clj-refactor.el
Comment (vemv) [inline clj-refactor.el:1979]: `seq-let` is probably nice but this lib (and I, coincidentally) only use seq.el for "seq stuff", not for other conveniences which would increase cognitive load to casual maintainers.

I'm generally happy with `let*` (and prefer `let*` over `let`, no matter what)

File: clj-refactor.el
Comment (vemv) [inline clj-refactor.el:2125]: Maybe add a `;; comment` explaining the case where `libspec` is nil

File: clj-refactor.el
Comment (vemv) [inline clj-refactor.el:1998]:
```suggestion
current point if it is within a reader conditional. If a given value is unknown, it will be expressed as nil"
```

or something to that effect. e.g. we never return nil, we always return a tuple

File: tests/unit-test.el
Comment (vemv) [inline tests/unit-test.el:177]: Let's use `<POINT>` as the marker, which although ugly, is unmistakable and will result in pretty self-describing tests IMO

File: tests/unit-test.el
Comment (vemv) [inline tests/unit-test.el:202]: We could exercise a few more cases:

* `?(:clj 1 :cljs (Math/log|)`
  * two branches present
* `?(:cljs (Math/log| :clj 1)`
  * two branches present, reverse order
* `?(|)`
  * incomplete reader conditional

File: tests/unit-test.el
Comment (vemv) [inline tests/unit-test.el:227]: Is it possible to add a counterexample where completing-read is called `> 0` times?

That would make this test more logically robust.

File: tests/unit-test.el
Comment (vemv) [inline tests/unit-test.el:239]: Let's set it to `<whatever the previous value was>` instead of `nil`

File: tests/unit-test.el
Comment (vemv) [inline tests/unit-test.el:250]: Fantastic tests, thank you 👏

**Review by vemv — APPROVED** [review]
This looks great to me. Thanks much!

I assume you've tried this over the real middleware we released for refactor-nrepl.

* Does it work correctly?
* Do you need this PR merged asap, or are you comfortable by working with your clj-refactor fork day-to-day?
* Do you have any new feedback item for the refactor-nrepl middleware, or shall we implement it as planned?

Cheers - V

File: clj-refactor.el
Comment (dgtized) [reply clj-refactor.el:1968]: Both `alias-ref` and `language-context` are described in the next paragraph. If you have suggestions on how to improve those descriptions for clarity I am happy to change them, but I need more details on the concern here.

File: clj-refactor.el
Comment (dgtized) [reply clj-refactor.el:2125]: I think this will be easier to follow once the `cljr-slash-uses-suggest-libspec` feature flag is gone. If we can't find a libspec we don't suggest or insert one, which is the same behavior we used previously but required two `when-let`. We could extract the case where `cljr-slash-uses-suggest-libspec` is nil into it's own function, maybe that would be clearer? Could you elaborate on what your concern is here, I don't think I understand.

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el:227]: That's what the other three tests are doing that are replacing the return value for 'completing-read with a `spy-on` with a `:and-return-value` mock return value. The reason this particular example is asserting it's not called is because it's one of the few cases it's not in the call graph. There may be a counter-example I missed though, so I would be happy to add one if I misunderstood, but I need more explanation.

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el:202]: I don't think the current method of parsing will handle all of that. I cribbed something together to support a few cases using the existing reader conditional code. However, this should also support, :cljr, :bb, and :default and the incomplete reader conditional, and all of those cause emacs to lockup in an infinite loop as the existing `cljr--point-in-reader-conditional-branch-p` method expects an exact match or loops forever. I'm not sure how critical this is just yet though, as from my understanding the middleware is still ignoring this field.

I've add the test cases here and mark the ones that are failing as pending but I think it may out of scope to address this in this PR as it's fixes for existing code. This should definitely be addressed before the flag is enabled by default though.

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el]: This is behavior is broken in buttercup anyway. It's supposed to support using :var ((cljr-slash-uses-suggest-libspec t)) as a lexical let around the entire describe but it just doesn't appear to work at present. I just swapped it with a local lexical let for each cljr-slash invocation to enable the feature for each call.

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el:177]: I cribbed this approach from `clojure-mode`, `cider`, and `smartparens` tests. I would much prefer to match a common formatting scheme over inventing our own. You are welcome to change how you see fit though.

Comment (vemv) [reply tests/unit-test.el:177]: > I cribbed this approach from clojure-mode, cider, and smartparens tests.

Ok, let's go with it.

File: tests/unit-test.el
Comment (vemv) [reply tests/unit-test.el:202]: > and all of those cause emacs to lockup in an infinite loop

Well, that seems an important issue, showing that it was a good use of code review. We most certainly don't want people experiencing/reporting infinite loops.

(Especially since this PR will use that faulty code more frequently than we do now)

It would be a much welcome side quest if you could fix this (here or separately - as you wish)

File: tests/unit-test.el
Comment (vemv) [reply tests/unit-test.el:266]: If, as a maintainer, I read `(expect 'completing-read :to-have-been-called-times 0)`, I will expect a counterexample which says `(expect 'completing-read :to-have-been-called-times <a positive number>)`, using the same construct, in a nearby test case.

Otherwise it is overassuming to expect people to detailedly understand all tests cases, and how test cases that use different constructs relate to each other.

Basically I'm suggesting to turn an implicit similarity into an explicit one.

File: tests/unit-test.el
Comment (vemv) [reply tests/unit-test.el]: Could you paste something similar as a `;; comment`?

Linking to a buttercup issue or such

Comment (dgtized) [reply tests/unit-test.el]: I don't have a buttercup issue for it, it just didn't work. I just changed it to use:

```
(let ((cljr-slash-uses-suggest-libspec t))
          (cljr-slash))
```

For those three test cases. That's correct though a little more verbose. If someone wants to clean it up or understands buttercup better so they can get it to work they are welcome to, I don't have the time to tweak it though, sorry.

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el]: Turns out I misread how the count expectations work, it has to be the at the end of the test. I thought it worked like rspec where it asserts at the beginning and then checks all assertions at the end. I added call count assertions at the end of all 4 of those test cases.

File: clj-refactor.el
Comment (dgtized) [reply clj-refactor.el]: The [emacs docs style guide](https://www.gnu.org/software/emacs/manual/html_node/elisp/Documentation-Tips.html) recommends 67 characters for the first line of a documentation so it looks better in apropos. . I was attempting to keep the documentation intro simple and explain it in more detail later to match that style recommendation.

I'm going to change it to "Suggest namespace require libspecs for an alias", and then expand the documentation below to include examples, crosslinking to `cljr--language-at-point` and the corresponding clojure docs. That way the complexity of the variables description is not introduced until the paragraph below. I hope that resolves your concern.

File: clj-refactor.el
Comment (dgtized) [reply clj-refactor.el:2138]: Couldn't think of a name to keep this clear, so just left a comment. Formatting was off anyway.

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el:291]: thanks!

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el:202]: I'm going to temporarily disable the reader conditional language context field while it's locking up and disable the corresponding tests for now. That way it's not introducing a higher risk of causing emacs to block, and we can re-enable it when the middleware starts supporting that field and implement a more robust parsing solution using the existing tests. I think the API is fine here though, we can improve it as we move forward.

File: clj-refactor.el
Comment (dgtized) [reply clj-refactor.el]: I've addressed this to your recommendation. From my perspective, destructuring binding is much easier to follow over manually parsing out car/cadr of each element, but to each their own. I appreciate your concern about cognitive load from unfamiliar constructs, but it's also frustrating and discouraging as a contributor to so frequently encounter code included in emacs core that is then dismissed at review. Do you have a style guide somewhere that I could follow to discover which are recommended and which are discouraged instead of discovering all of these at review time?

File: tests/unit-test.el
Comment (dgtized) [reply tests/unit-test.el:203]: Extracted into https://github.com/clojure-emacs/clj-refactor.el/issues/533

**Comment by dgtized** [comment]
I believe I've addressed all outstanding concerns, updated external documentation and added a changelog entry as well as manually retesting it in a project. I would much prefer we merge this over letting it bitrot in review. It's not perfect but it will allow folks to start testing and working off of the new middleware so we can iterate from there. I believe the next steps beyond this PR on the cljr-refactor side are to address the reader conditional parsing in #533, and then document the new syntax for `cljr-magic-require-namespaces` in https://github.com/clojure-emacs/clj-refactor.el/issues/530, though I think all of the logic there is in the middleware, so it's probably better to document that after it's implemented.

I believe that API is solid enough to move forward with. I didn't encounter anything critical with it in development, but now that the logic has moved into the middleware, it's likely we won't see issues in the recommendations until we encounter them in actual use.

Let me know if there is any other critical concerns that need to be addressed prior to merge.

**Comment by dgtized** [comment]
Just checking back in to see what else needs to be done here before it's merged.

**Comment by vemv** [comment]
Thanks for the ping! Will check out today / tomorrow at most

File: clj-refactor.el
Comment (vemv) [reply clj-refactor.el]: >  I could follow to discover which are recommended and which are discouraged instead of discovering all of these at review time?

Not that I know of

Generally, afaict in clojure-emacs .el code we lean toward simple constructs. Even then, different reviewers might have different preferences.

My genuine recommendation would be to simply be ready to tweak some things upon request. I myself cannot predict the feedback I'll receive in the variety of repos I contibute to - some little churn on superficial things is just part of life! :)

**Comment by vemv** [comment]
Thanks again for this first step!

Should https://github.com/clojure-emacs/refactor-nrepl/issues/384 be implemented as planned? Anything that came to mind more recently?

**Comment by dgtized** [comment]
I think we can proceed with the remainder of clojure-emacs/refactor-nrepl#384 as planned. However, until #533 is implemented, we won't know the correct current language context at point. My expectation remains that we will have more tweaking required on the exact recommendations from the backend, but that won't be as visible until it's in use. I'll see if I can take a stab at #533 in the coming week or two, though so it can operate with the correct local information.

Thanks for merging this, and the review!

**Comment by vemv** [comment]
Nice! I'll try to advance on the middleware, do feel free to ping if you'd need to have it sooner.

**Pull request merged**
